File: apps/api/src/alerts/alert.repository.ts
import { Injectable } from '@nestjs/common';
import {
  AlertChannel,
  AlertChannelType,
  AlertDeliveryStatus,
  AlertEventKind,
  AlertRule,
  AlertSeverity,
  AlertTrigger,
  Prisma,
} from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { CryptoService } from '../common/crypto/crypto.service';
import type { RuleScope } from './alerts.types';

type ChannelView = Omit<AlertChannel, 'secretEnc'>;
const redactChannel = (c: AlertChannel): ChannelView => {
  const { secretEnc: _omit, ...rest } = c;
  return rest;
};

@Injectable()
export class AlertRepository {
  constructor(
    private readonly prisma: PrismaService,
    private readonly crypto: CryptoService,
  ) {}

  // ── channels ──
  async createChannel(
    orgId: string,
    dto: {
      type: AlertChannelType;
      name: string;
      enabled?: boolean;
      config?: Prisma.JsonValue;
      secret?: string | null;
    },
  ): Promise<ChannelView> {
    const c = await this.prisma.alertChannel.create({
      data: {
        organizationId: orgId,
        type: dto.type,
        name: dto.name,
        enabled: dto.enabled ?? true,
        config: (dto.config ?? {}) as Prisma.InputJsonValue,
        secretEnc: dto.secret ? this.crypto.encrypt(dto.secret) : null,
      },
    });
    return redactChannel(c);
  }
  async listChannels(orgId: string): Promise<ChannelView[]> {
    return (
      await this.prisma.alertChannel.findMany({
        where: { organizationId: orgId },
        orderBy: { name: 'asc' },
      })
    ).map(redactChannel);
  }
  async getChannel(orgId: string, id: string): Promise<ChannelView | null> {
    const c = await this.prisma.alertChannel.findFirst({ where: { id, organizationId: orgId } });
    return c ? redactChannel(c) : null;
  }
  async deleteChannel(orgId: string, id: string): Promise<void> {
    await this.prisma.alertChannel.deleteMany({ where: { id, organizationId: orgId } });
  }
  /** Full rows (WITH secretEnc) — for the delivery path only. */
  async channelsByIds(ids: string[]): Promise<AlertChannel[]> {
    if (!ids.length) return [];
    return this.prisma.alertChannel.findMany({ where: { id: { in: ids } } });
  }

  // ── rules ──
  async createRule(
    orgId: string,
    dto: {
      name: string;
      trigger: AlertTrigger;
      scope: RuleScope;
      enabled?: boolean;
      targetStates?: string[];
      metric?: string | null;
      op?: string | null;
      threshold?: number | null;
      forSeconds?: number | null;
      severity: AlertSeverity;
      channelIds: string[];
      cooldownSeconds: number;
      notifyOnRecovery: boolean;
    },
  ): Promise<AlertRule> {
    return this.prisma.alertRule.create({
      data: {
        organizationId: orgId,
        name: dto.name,
        trigger: dto.trigger,
        scope: dto.scope as unknown as Prisma.InputJsonValue,
        enabled: dto.enabled ?? true,
        targetStates: dto.targetStates ?? [],
        metric: dto.metric ?? null,
        op: dto.op ?? null,
        threshold: dto.threshold ?? null,
        forSeconds: dto.forSeconds ?? null,
        severity: dto.severity,
        channelIds: dto.channelIds,
        cooldownSeconds: dto.cooldownSeconds,
        notifyOnRecovery: dto.notifyOnRecovery,
      },
    });
  }
  listRules(orgId: string): Promise<AlertRule[]> {
    return this.prisma.alertRule.findMany({ where: { organizationId: orgId }, orderBy: { name: 'asc' } });
  }
  getRule(orgId: string, id: string): Promise<AlertRule | null> {
    return this.prisma.alertRule.findFirst({ where: { id, organizationId: orgId } });
  }
  async deleteRule(orgId: string, id: string): Promise<void> {
    await this.prisma.alertRule.deleteMany({ where: { id, organizationId: orgId } });
  }
  updateRule(
    orgId: string,
    id: string,
    patch: Partial<{ enabled: boolean; name: string }>,
  ): Promise<Prisma.BatchPayload> {
    return this.prisma.alertRule.updateMany({ where: { id, organizationId: orgId }, data: patch });
  }
  enabledRules(orgId: string, trigger: AlertTrigger): Promise<AlertRule[]> {
    return this.prisma.alertRule.findMany({ where: { organizationId: orgId, enabled: true, trigger } });
  }
  allEnabledRules(trigger: AlertTrigger): Promise<AlertRule[]> {
    return this.prisma.alertRule.findMany({ where: { enabled: true, trigger } });
  }
  /** The device's networkId (for networkIds-scoped rules); null if the device is absent. */
  async deviceNetworkId(orgId: string, deviceId: string): Promise<string | null> {
    const d = await this.prisma.device.findFirst({
      where: { id: deviceId, organizationId: orgId }, select: { networkId: true },
    });
    return d?.networkId ?? null;
  }

  // ── events + deliveries ──
  createEvent(e: {
    organizationId: string;
    ruleId: string;
    deviceId: string | null;
    kind: AlertEventKind;
    severity: AlertSeverity;
    detail: Prisma.InputJsonValue;
    dedupKey: string;
  }) {
    return this.prisma.alertEvent.create({ data: e });
  }
  latestEventFor(orgId: string, ruleId: string, deviceId: string | null) {
    return this.prisma.alertEvent.findFirst({
      where: { organizationId: orgId, ruleId, deviceId },
      orderBy: { createdAt: 'desc' },
    });
  }
  listEvents(orgId: string, limit = 100) {
    return this.prisma.alertEvent.findMany({
      where: { organizationId: orgId },
      orderBy: { createdAt: 'desc' },
      take: limit,
    });
  }
  async enqueueDeliveries(eventId: string, channelIds: string[], now: Date): Promise<void> {
    if (!channelIds.length) return;
    await this.prisma.alertDelivery.createMany({
      data: channelIds.map((channelId) => ({ alertEventId: eventId, channelId, nextAttemptAt: now })),
    });
  }
  dueDeliveries(now: Date, limit = 50) {
    return this.prisma.alertDelivery.findMany({
      where: { status: { in: ['PENDING', 'FAILED'] }, nextAttemptAt: { lte: now } },
      orderBy: { nextAttemptAt: 'asc' },
      take: limit,
      include: { event: true },
    });
  }
  async markDelivery(
    id: string,
    patch: {
      status: AlertDeliveryStatus;
      attempts?: number;
      nextAttemptAt?: Date;
      lastError?: string | null;
      lastAttemptAt?: Date;
    },
  ): Promise<void> {
    await this.prisma.alertDelivery.update({ where: { id }, data: patch });
  }
}

File: src/Modules/Alerting/NodeScope.Modules.Alerting.Infrastructure/Persistence/AlertRepository.cs
using System.Data;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using NodeScope.Modules.Alerting.Application;
using NodeScope.Modules.Alerting.Domain;
using NodeScope.Platform.Abstractions;
using Npgsql;

namespace NodeScope.Modules.Alerting.Infrastructure.Persistence;

internal sealed class AlertRepository : IAlertRepository
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);
    private readonly AlertingDbContext _db;

    public AlertRepository(AlertingDbContext db)
    {
        _db = db;
    }

    public async Task<AlertChannelRecord> CreateChannelAsync(
        NewAlertChannel channel,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(channel);
        var now = DateTime.UtcNow;
        var row = new AlertChannelRow
        {
            Id = Guid.NewGuid().ToString(),
            OrganizationId = channel.OrganizationId,
            Type = channel.Type,
            Name = channel.Name,
            Enabled = channel.Enabled,
            ConfigJson = channel.ConfigJson,
            SecretEnc = channel.SecretEnc,
            Version = 1,
            CreatedAt = now,
            UpdatedAt = now,
        };
        _db.Channels.Add(row);
        await SaveUniqueAsync("ALERT_005", "CHANNEL_NAME_ALREADY_EXISTS", cancellationToken);
        return ToRecord(row);
    }

    public async Task<IReadOnlyList<AlertChannelRecord>> ListChannelsAsync(
        string organizationId,
        CancellationToken cancellationToken) =>
        [.. (await _db.Channels.AsNoTracking()
                .Where(row => row.OrganizationId == organizationId)
                .OrderBy(row => row.CreatedAt)
                .ToListAsync(cancellationToken))
            .Select(ToRecord)];

    public async Task<AlertChannelRecord?> FindChannelAsync(
        string organizationId,
        string channelId,
        CancellationToken cancellationToken)
    {
        var row = await _db.Channels.AsNoTracking()
            .SingleOrDefaultAsync(
                channel => channel.Id == channelId && channel.OrganizationId == organizationId,
                cancellationToken);
        return row is null ? null : ToRecord(row);
    }

    public async Task<bool> DeleteChannelAsync(
        string organizationId,
        string channelId,
        CancellationToken cancellationToken)
    {
        var row = await _db.Channels.SingleOrDefaultAsync(
            channel => channel.Id == channelId && channel.OrganizationId == organizationId,
            cancellationToken);
        if (row is null)
        {
            return false;
        }

        if (await _db.RuleChannels.AnyAsync(link => link.ChannelId == channelId, cancellationToken))
        {
            throw new ApiException("ALERT_002", "CHANNEL_IN_USE", 409);
        }

        _db.Channels.Remove(row);
        try
        {
            await _db.SaveChangesAsync(cancellationToken);
        }
        catch (DbUpdateException exception) when (
            exception.InnerException is PostgresException
            {
                SqlState: PostgresErrorCodes.ForeignKeyViolation,
            })
        {
            throw new ApiException("ALERT_002", "CHANNEL_IN_USE", 409);
        }

        return true;
    }

    public async Task<AlertRuleRecord> CreateRuleAsync(
        NewAlertRule rule,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(rule);
        var now = DateTime.UtcNow;
        var row = new AlertRuleRow
        {
            Id = Guid.NewGuid().ToString(),
            OrganizationId = rule.OrganizationId,
            Name = rule.Name,
            Enabled = rule.Enabled,
            Trigger = rule.Trigger,
            ScopeJson = JsonSerializer.Serialize(AlertScopeDto.From(rule.Scope), JsonOptions),
            TargetStates = [.. rule.TargetStates],
            Metric = rule.Metric,
            Op = rule.Op,
            Threshold = rule.Threshold,
            ForSeconds = rule.ForSeconds,
            Severity = rule.Severity,
            CooldownSeconds = rule.CooldownSeconds,
            NotifyOnRecovery = rule.NotifyOnRecovery,
            Version = 1,
            CreatedAt = now,
            UpdatedAt = now,
        };
        foreach (var channelId in rule.ChannelIds)
        {
            row.Channels.Add(new AlertRuleChannelRow { RuleId = row.Id, ChannelId = channelId });
        }

        _db.Rules.Add(row);
        try
        {
            await _db.SaveChangesAsync(cancellationToken);
        }
        catch (DbUpdateException exception) when (
            exception.InnerException is PostgresException
            {
                SqlState: PostgresErrorCodes.UniqueViolation,
            })
        {
            throw new ApiException("ALERT_006", "RULE_NAME_ALREADY_EXISTS", 409);
        }
        catch (DbUpdateException exception) when (
            exception.InnerException is PostgresException
            {
                SqlState: PostgresErrorCodes.ForeignKeyViolation,
            })
        {
            throw new ApiException("ALERT_001", "CHANNEL_NOT_FOUND", 400);
        }

        return ToRecord(row);
    }

    public async Task<IReadOnlyList<AlertRuleRecord>> ListRulesAsync(
        string organizationId,
        CancellationToken cancellationToken) =>
        [.. (await _db.Rules.AsNoTracking()
                .Include(rule => rule.Channels)
                .Where(rule => rule.OrganizationId == organizationId)
                .OrderBy(rule => rule.CreatedAt)
                .ToListAsync(cancellationToken))
            .Select(ToRecord)];

    public async Task<IReadOnlyList<AlertRuleRecord>> ListEnabledRulesAsync(
        AlertTrigger trigger,
        string? organizationId,
        CancellationToken cancellationToken)
    {
        var query = _db.Rules.AsNoTracking()
            .Include(rule => rule.Channels)
            .Where(rule => rule.Enabled && rule.Trigger == trigger);
        if (organizationId is not null)
        {
            query = query.Where(rule => rule.OrganizationId == organizationId);
        }

        return [.. (await query.OrderBy(rule => rule.Id).ToListAsync(cancellationToken)).Select(ToRecord)];
    }

    public async Task<bool> DeleteRuleAsync(
        string organizationId,
        string ruleId,
        CancellationToken cancellationToken)
    {
        var row = await _db.Rules.SingleOrDefaultAsync(
            rule => rule.Id == ruleId && rule.OrganizationId == organizationId,
            cancellationToken);
        if (row is null)
        {
            return false;
        }

        _db.Rules.Remove(row);
        await _db.SaveChangesAsync(cancellationToken);
        return true;
    }

    public async Task<IReadOnlyList<AlertEventRecord>> ListEventsAsync(
        string organizationId,
        int take,
        CancellationToken cancellationToken) =>
        [.. (await _db.Events.AsNoTracking()
                .Where(alertEvent => alertEvent.OrganizationId == organizationId)
                .OrderByDescending(alertEvent => alertEvent.CreatedAt)
                .ThenByDescending(alertEvent => alertEvent.Id)
                .Take(take)
                .ToListAsync(cancellationToken))
            .Select(ToRecord)];

    public async Task<AlertEventRecord?> RecordTransitionAsync(
        AlertRuleRecord rule,
        string deviceId,
        AlertEventKind kind,
        string detailJson,
        DateTime nowUtc,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(rule);
        var dedupKey = $"{rule.Id}:{deviceId}";
        await using var transaction = await _db.Database.BeginTransactionAsync(
            IsolationLevel.ReadCommitted,
            cancellationToken);
        await _db.Database.ExecuteSqlInterpolatedAsync(
            $"SELECT pg_advisory_xact_lock(hashtextextended({dedupKey}, 0))",
            cancellationToken);

        var incident = await _db.Incidents.SingleOrDefaultAsync(
            row => row.DedupKey == dedupKey,
            cancellationToken);
        if (kind == AlertEventKind.Firing)
        {
            if (incident?.IsOpen == true
                || incident?.LastFiredAt is { } lastFired
                && nowUtc < lastFired.AddSeconds(rule.CooldownSeconds))
            {
                await transaction.CommitAsync(cancellationToken);
                return null;
            }

            if (incident is null)
            {
                incident = new AlertIncidentRow
                {
                    DedupKey = dedupKey,
                    OrganizationId = rule.OrganizationId,
                    RuleId = rule.Id,
                    DeviceId = deviceId,
                };
                _db.Incidents.Add(incident);
            }

            incident.IsOpen = true;
            incident.LastFiredAt = nowUtc;
            incident.UpdatedAt = nowUtc;
        }
        else
        {
            if (incident?.IsOpen != true)
            {
                await transaction.CommitAsync(cancellationToken);
                return null;
            }

            incident.IsOpen = false;
            incident.LastResolvedAt = nowUtc;
            incident.UpdatedAt = nowUtc;
        }

        var eventRow = new AlertEventRow
        {
            Id = Guid.NewGuid().ToString(),
            OrganizationId = rule.OrganizationId,
            RuleId = rule.Id,
            RuleName = rule.Name,
            DeviceId = deviceId,
            Kind = kind,
            Severity = rule.Severity,
            DetailJson = detailJson,
            DedupKey = dedupKey,
            CreatedAt = nowUtc,
        };
        _db.Events.Add(eventRow);
        foreach (var channelId in rule.ChannelIds)
        {
            _db.Deliveries.Add(new AlertDeliveryRow
            {
                Id = Guid.NewGuid().ToString(),
                AlertEventId = eventRow.Id,
                ChannelId = channelId,
                Status = AlertDeliveryStatus.Pending,
                Attempts = 0,
                NextAttemptAt = nowUtc,
                CreatedAt = nowUtc,
            });
        }

        await _db.SaveChangesAsync(cancellationToken);
        await transaction.CommitAsync(cancellationToken);
        return ToRecord(eventRow);
    }

    public async Task<IReadOnlyList<MetricAggregate>> QuerySustainedMetricAsync(
        string organizationId,
        string metric,
        string op,
        IReadOnlyList<string>? deviceIds,
        DateTime sinceUtc,
        DateTime nowUtc,
        CancellationToken cancellationToken)
    {
        var window = nowUtc - sinceUtc;
        var boundaryFloorUtc = sinceUtc - window;
        var query = deviceIds is null
            ? _db.Database.SqlQuery<MetricAggregate>(
                $"""
                WITH "BoundarySamples" AS (
                    SELECT DISTINCT ON ("deviceId") "deviceId", "value", "time"
                    FROM "MonitoringMetric"
                    WHERE "organizationId" = {organizationId}
                      AND "metric" = {metric}
                      AND "time" >= {boundaryFloorUtc}
                      AND "time" < {sinceUtc}
                    ORDER BY "deviceId", "time" DESC
                ),
                "WindowSamples" AS (
                    SELECT "deviceId", "value", "time"
                    FROM "MonitoringMetric"
                    WHERE "organizationId" = {organizationId}
                      AND "metric" = {metric}
                      AND "time" >= {sinceUtc}
                      AND "time" <= {nowUtc}
                ),
                "Samples" AS (
                    SELECT * FROM "BoundarySamples"
                    UNION ALL
                    SELECT * FROM "WindowSamples"
                )
                SELECT "deviceId" AS "DeviceId",
                       CASE WHEN {op} = 'lt' THEN MAX("value") ELSE MIN("value") END AS "Value"
                FROM "Samples"
                GROUP BY "deviceId"
                HAVING COUNT(*) FILTER (WHERE "time" < {sinceUtc}) > 0
                   AND COUNT(*) FILTER (WHERE "time" >= {sinceUtc}) > 0
                """)
            : _db.Database.SqlQuery<MetricAggregate>(
                $"""
                WITH "BoundarySamples" AS (
                    SELECT DISTINCT ON ("deviceId") "deviceId", "value", "time"
                    FROM "MonitoringMetric"
                    WHERE "organizationId" = {organizationId}
                      AND "metric" = {metric}
                      AND "time" >= {boundaryFloorUtc}
                      AND "time" < {sinceUtc}
                      AND "deviceId" = ANY ({deviceIds.ToArray()})
                    ORDER BY "deviceId", "time" DESC
                ),
                "WindowSamples" AS (
                    SELECT "deviceId", "value", "time"
                    FROM "MonitoringMetric"
                    WHERE "organizationId" = {organizationId}
                      AND "metric" = {metric}
                      AND "time" >= {sinceUtc}
                      AND "time" <= {nowUtc}
                      AND "deviceId" = ANY ({deviceIds.ToArray()})
                ),
                "Samples" AS (
                    SELECT * FROM "BoundarySamples"
                    UNION ALL
                    SELECT * FROM "WindowSamples"
                )
                SELECT "deviceId" AS "DeviceId",
                       CASE WHEN {op} = 'lt' THEN MAX("value") ELSE MIN("value") END AS "Value"
                FROM "Samples"
                GROUP BY "deviceId"
                HAVING COUNT(*) FILTER (WHERE "time" < {sinceUtc}) > 0
                   AND COUNT(*) FILTER (WHERE "time" >= {sinceUtc}) > 0
                """);

        return await query.ToListAsync(cancellationToken);
    }

    public async Task<IReadOnlyList<AlertDeliveryRecord>> ListDueDeliveriesAsync(
        DateTime nowUtc,
        int take,
        CancellationToken cancellationToken) =>
        [.. (await _db.Deliveries.AsNoTracking()
                .Include(delivery => delivery.Event)
                .Where(delivery =>
                    (delivery.Status == AlertDeliveryStatus.Pending
                     || delivery.Status == AlertDeliveryStatus.Failed)
                    && delivery.NextAttemptAt <= nowUtc)
                .OrderBy(delivery => delivery.NextAttemptAt)
                .ThenBy(delivery => delivery.Id)
                .Take(take)
                .ToListAsync(cancellationToken))
            .Select(ToRecord)];

    public async Task<IReadOnlyList<AlertChannelRecord>> FindChannelsAsync(
        IReadOnlyCollection<string> channelIds,
        CancellationToken cancellationToken)
    {
        if (channelIds.Count == 0)
        {
            return [];
        }

        return [.. (await _db.Channels.AsNoTracking()
                .Where(channel => channelIds.Contains(channel.Id))
                .ToListAsync(cancellationToken))
            .Select(ToRecord)];
    }

    public async Task UpdateDeliveryAsync(
        string deliveryId,
        AlertDeliveryStatus status,
        int attempts,
        DateTime? lastAttemptAt,
        DateTime? nextAttemptAt,
        string? lastError,
        CancellationToken cancellationToken)
    {
        var row = await _db.Deliveries.SingleAsync(
            delivery => delivery.Id == deliveryId,
            cancellationToken);
        row.Status = status;
        row.Attempts = attempts;
        row.LastAttemptAt = lastAttemptAt;
        if (nextAttemptAt is not null)
        {
            row.NextAttemptAt = nextAttemptAt.Value;
        }

        row.LastError = lastError;
        await _db.SaveChangesAsync(cancellationToken);
    }

    public async Task<(int Up, int Down, int Warning, int Unknown)> CountDeviceStatesAsync(
        CancellationToken cancellationToken)
    {
        var rows = await _db.Database.SqlQueryRaw<StateCount>(
                """
                SELECT "state"::text AS "State", COUNT(*)::int AS "Count"
                FROM "DeviceStatus"
                GROUP BY "state"
                """)
            .ToListAsync(cancellationToken);
        var counts = rows.ToDictionary(row => row.State, row => row.Count, StringComparer.Ordinal);
        return (
            counts.GetValueOrDefault("UP"),
            counts.GetValueOrDefault("DOWN"),
            counts.GetValueOrDefault("WARNING"),
            counts.GetValueOrDefault("UNKNOWN"));
    }

    private async Task SaveUniqueAsync(
        string code,
        string message,
        CancellationToken cancellationToken)
    {
        try
        {
            await _db.SaveChangesAsync(cancellationToken);
        }
        catch (DbUpdateException exception) when (
            exception.InnerException is PostgresException
            {
                SqlState: PostgresErrorCodes.UniqueViolation,
            })
        {
            throw new ApiException(code, message, 409);
        }
    }

    private static AlertChannelRecord ToRecord(AlertChannelRow row) => new(
        row.Id,
        row.OrganizationId,
        row.Type,
        row.Name,
        row.Enabled,
        row.ConfigJson,
        row.SecretEnc,
        row.Version,
        row.CreatedAt,
        row.UpdatedAt);

    private static AlertRuleRecord ToRecord(AlertRuleRow row)
    {
        var dto = JsonSerializer.Deserialize<AlertScopeDto>(row.ScopeJson, JsonOptions)
            ?? throw new InvalidOperationException($"Rule {row.Id} has an invalid scope");
        var errors = new List<string>();
        var scope = dto.Validate(errors);
        if (scope is null || errors.Count > 0)
        {
            throw new InvalidOperationException($"Rule {row.Id} has an invalid scope");
        }

        return new AlertRuleRecord(
            row.Id,
            row.OrganizationId,
            row.Name,
            row.Enabled,
            row.Trigger,
            scope,
            row.TargetStates,
            row.Metric,
            row.Op,
            row.Threshold,
            row.ForSeconds,
            row.Severity,
            [.. row.Channels.Select(link => link.ChannelId).Order(StringComparer.Ordinal)],
            row.CooldownSeconds,
            row.NotifyOnRecovery,
            row.Version,
            row.CreatedAt,
            row.UpdatedAt);
    }

    private static AlertEventRecord ToRecord(AlertEventRow row) => new(
        row.Id,
        row.OrganizationId,
        row.RuleId,
        row.RuleName,
        row.DeviceId,
        row.Kind,
        row.Severity,
        row.DetailJson,
        row.DedupKey,
        row.CreatedAt);

    private static AlertDeliveryRecord ToRecord(AlertDeliveryRow row) => new(
        row.Id,
        row.AlertEventId,
        row.ChannelId,
        row.Status,
        row.Attempts,
        row.LastAttemptAt,
        row.NextAttemptAt,
        row.LastError,
        ToRecord(row.Event));

    private sealed record StateCount(string State, int Count);
}
